lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json/JsonDeserializer.java

```diff
@@ -0,0 +1,838 @@
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            DATE.writeLong(builder, toIntExact(parseHiveDate(parser.getText()).toEpochDay()));
+        }
+    }
+
+    private static class TimestampDecoder
+            extends Decoder
+    {
+        private final TimestampType timestampType;
+        private final Function<String, DecodedTimestamp> timestampParser;
+
+        public TimestampDecoder(TimestampType timestampType, Function<String, DecodedTimestamp> timestampParser)
+        {
+            super(timestampType);
+            this.timestampType = timestampType;
+            this.timestampParser = timestampParser;
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            DecodedTimestamp timestamp = timestampParser.apply(parser.getText());
+            createTimestampEncoder(timestampType, UTC).write(timestamp, builder);
+        }
+    }
+
+    private static class VarbinaryDecoder
+            extends Decoder
+    {
+        // This must be created per instance because it is not thread safe
+        private final CharsetDecoder charsetDecoder = createCharsetDecoder();
+
+        public VarbinaryDecoder()
+        {
+            super(VARBINARY);
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            VARBINARY.writeSlice(builder, parseBinary(parser.getText(), charsetDecoder));
+        }
+
+        private static Slice parseBinary(String value, CharsetDecoder charsetDecoder1)
+                throws IOException
+        {
+            byte[] utf8 = value.getBytes(UTF_8);
+            // This corrupts the data, but this is exactly what Hive does, so we get the same result as Hive
+            String decode = charsetDecoder1.decode(ByteBuffer.wrap(utf8, 0, utf8.length)).toString();
+            return Slices.utf8Slice(decode);
+        }
+
+        private static CharsetDecoder createCharsetDecoder()
+        {
+            return UTF_8.newDecoder()
+                    .onMalformedInput(CodingErrorAction.REPLACE)
+                    .onUnmappableCharacter(CodingErrorAction.REPLACE);
+        }
+    }
+
+    private static class VarcharDecoder
+            extends Decoder
+    {
+        private final VarcharType varcharType;
+
+        public VarcharDecoder(VarcharType varcharType)
+        {
+            super(varcharType);
+            this.varcharType = varcharType;
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            varcharType.writeSlice(builder, truncateToLength(Slices.utf8Slice(parser.getText()), varcharType));
```

Comment:
Do other readers truncate when the value doesn't fit in the declared length?

Reply:
They are supposed to and the generic one does this.